ansible_sdk/_util/metrics.py

```diff
@@ -0,0 +1,207 @@
+@dataclasses.dataclass(frozen=True, kw_only=True)
+class AnsibleRoleStats:
+    job_id: str
+    role_fqcn: str
+    collection_version: str
+    task_count: int
+    task_duration: str
+
+
+@dataclasses.dataclass(frozen=False, kw_only=True)
+class MetricsData:
+    started: datetime = dataclasses.field(default_factory=datetime.now)
+    task_counter: int = 0
+    task_data: dict = dataclasses.field(default_factory=dict)
+    role_data: dict = dataclasses.field(default_factory=dict)
+    collection_data: dict = dataclasses.field(default_factory=dict)
+
+
+class MetricsCalc:
+    def __init__(self):
+        pass
+
+    async def collect_metrics(self, status_obj: AnsibleJobStatus):
+        metrics_data = MetricsData()
+        job_csv_filename = os.path.join(
+            status_obj._job_def.metrics_output_path, "jobs.csv"
+        )
+        module_csv_filename = os.path.join(
+            status_obj._job_def.metrics_output_path, "modules.csv",
+        )
+        collection_csv_filename = os.path.join(
+            status_obj._job_def.metrics_output_path, "collections.csv"
+        )
+        roles_csv_filename = os.path.join(
+            status_obj._job_def.metrics_output_path, "roles.csv"
+        )
+
+        job_headers = [x.name for x in dataclasses.fields(AnsibleJobStats)]
+        module_headers = [x.name for x in dataclasses.fields(AnsibleModuleStats)]
+        collection_headers = [x.name for x in dataclasses.fields(AnsibleCollectionStats)]
+        role_headers = [x.name for x in dataclasses.fields(AnsibleRoleStats)]
+
+        with open(job_csv_filename, "w") as job_csv_fh, \
+                open(collection_csv_filename, "w") as collection_csv_fh, \
+                open(roles_csv_filename, "w") as role_csv_fh, \
+                open(module_csv_filename, "w") as module_csv_fh:
+            job_csv_writer = CSVAsync(job_csv_fh, job_headers, restval="NULL")
+            await job_csv_writer.writeheader_async()
+            module_csv_writer = CSVAsync(module_csv_fh, module_headers, restval="NULL")
+            await module_csv_writer.writeheader_async()
+            collection_csv_writer = CSVAsync(
+                collection_csv_fh, collection_headers, restval="NULL"
+            )
+            await collection_csv_writer.writeheader_async()
+            role_csv_writer = CSVAsync(role_csv_fh, role_headers, restval="NULL")
+            await role_csv_writer.writeheader_async()
+
+            async for ev in status_obj.events:
+                runner_ident = ev.get("runner_ident")
+                end_time = datetime.fromisoformat(ev.get("created"))
+                if ev.get("event") == "playbook_on_start":
+                    metrics_data.started = datetime.fromisoformat(ev.get("created"))
+                if ev.get("event") == "playbook_on_task_start":
+                    metrics_data.task_counter += 1
+                    # Update the count of task
+                    resolved_task_name = ev["event_data"]["resolved_action"]
```

Comment:
All this code needs to be very defensive about malformed events- unhandled exceptions in here will still bubble out to user code, so we need to be very careful about always accessing event dictionaries with `get()` and skipping events that don't have the "shape" we expect.